src/storage/block_store.cpp

```diff
@@ -517,3 +534,309 @@ void BlockStore::ScheduleTask() {
+    if (height > 1 && !SaveHeadHeight(height - 1)) {
+        spdlog::error("Failed to update head height");
+        return false;
+    }
+
+    return true;
+}
+
+bool BlockStore::DeleteDBMs(uint64_t height) {
+    uint64_t currentHeight = GetHeadHeight();
+    spdlog::info("Start to delete db ms record from {} to {}", height, currentHeight);
+    for (uint64_t h = height; h <= currentHeight; h++) {
+        auto res = dbStore_.DeleteMsPos(h);
+        if (res) {
+            spdlog::debug("Deleted Ms Pos at height {}", h);
+        } else {
+            spdlog::error("Failed to delete Ms Pos at height {}, DB record is not consistent", h);
+            return false;
+        }
+    }
+    return true;
+}
+
+bool BlockStore::DeleteDBBlks(uint64_t height) {
+    return dbStore_.DeleteBatchVtxPos(height);
+}
+
+bool BlockStore::RebuildUTXOAndReg(uint64_t height) {
+    // delete two columns in db  UTXO, Reg
+    std::string column1 = "utxo";
+    std::string column2 = "reg";
+    auto status1        = dbStore_.DeleteColumn(column1);
+    auto status2        = dbStore_.DeleteColumn(column2);
+    if (!status1 || !status2) {
+        return false;
+    }
+
+    // create the two new columns
+    auto statusU = dbStore_.CreateColumn(column1);
+    auto statusR = dbStore_.CreateColumn(column2);
+    if (!statusU || !statusR) {
+        return false;
+    }
+
+    for (uint64_t h = 0; h < height; h++) {
+        auto levelset = GetLevelSetVtcsAt(h, true);
+        std::swap(levelset.front(), levelset.back());
+        if (!ConstructUTXOAndRegFromLvs(levelset)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool BlockStore::ConstructUTXOAndRegFromLvs(std::vector<VertexPtr>& levelset) {
+    for (auto& vertex : levelset) {
+        if (ConstructUTXOAndRegFromVtx(vertex)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool BlockStore::ConstructUTXOAndRegFromVtx(const VertexPtr& vtx) {
+    size_t size   = vtx->cblock->GetTransactionSize();
+    auto validity = vtx->validity;
+    auto txs      = vtx->cblock->GetTransactions();
+
+    auto blkHash  = vtx->cblock->GetHash();
+    auto prevHash = vtx->cblock->GetPrevHash();
+
+    RegChange regChange;
+    TXOC txoc;
+    std::vector<UTXOPtr> newUXTOs;
+    if (vtx->cblock->IsFirstRegistration()) {
+        regChange.Create(blkHash, blkHash);
+    } else {
+        for (int txIndex = 0; txIndex < size; ++txIndex) {
+            if (validity[txIndex] == Vertex::VALID) {
+                auto tx = vtx->cblock->GetTransactions()[txIndex];
+
+                // input, find spent utxo and new reg
+                for (auto& input : tx->GetInputs()) {
+                    auto oldRedempHash = STORE->GetPrevRedemHash(prevHash);
+                    if (oldRedempHash.IsNull()) {
+                        spdlog::error("Can't find redemption hash {}", oldRedempHash.GetHex());
+                        return false;
+                    }
```

Comment:
The `regChange` is not flushed into DB yet so this can happen when there are more than one reg on a peer chain in the same level set, but this is not an error.